### ISSUE TITLE
mssql-conf settings ignored if file is copied to container with Visual Studio default UTF-8 W/BOM encoding

### ISSUE THREAD

**Issue by abair34** [microsoft/mssql-docker#586]
I have a project in VS 2019 that uses a SQL 2019 docker container that requires trace flags to be enabled.  To enable the trace flags I copy a mssql.conf file into the docker image through the dockerfile.  Unfortunately it seems trace flags are not able to be configured through environment variables.

When copying a mssql.conf file into the container from visual studio all settings were being ignored.  

I tracked down the issue to the mssql.conf file's encoding which was built with Visual Studio's default encoding for text files.  It would be a nice improvement if mssql-conf could handle UTF-8 with signature 

The text file was saved as "Unicode (UTF-8 with signature) - Codepage 65001"

**Workaround:**
- In visual studio when saving mssql.conf use file->save as
- click the down arrow on the save button and click save with encoding
- click yes to replace
- select Unicode (UTF-8 without signature) - Codepage 65001 (this is far down at the bottom, remember the key is to select the one WITHOUT signature)

**Repro**

DockerFile
```
FROM mcr.microsoft.com/mssql/server:2019-latest
ENV MSSQL_SA_PASSWORD="Pass@word"
ENV ACCEPT_EULA="Y"
ENV MSSQL_PID="Developer"
ENV MSSQL_AGENT_ENABLED="True"
COPY ["SQLLinux/MSSQLServer/mssql.conf", "/var/opt/mssql/"]

```
The following mssql.conf file is in the SQLLinux\MSSQLServer folder as referenced in the above conf statement

```
[traceflag]
traceflag0 = 9481
traceflag1 = 2371
```

Expected:
When container is built, the mssql.conf file will get copied into the linux container and will run with the 2 trace flags enabled.

`/opt/mssql/bin/mssql-conf validate `
should return 
> Validation successful.

Actual:
SQL starts but the trace flags are not enabled.  

`/opt/mssql/bin/mssql-conf validate`
returns

> There was a parsing error in the configuration file.

`cat -A /var/opt/mssql/mssql.conf `
returned 

> M-oM-;M-?[traceflag]$
> traceflag0 = 9481$
> traceflag1 = 2371$

**Potential Fix for python 2.7**
/opt/mssql/lib/mssql-conf/mssqlconfhelper.py
`import codecs
....
Line 603 
           config.read(configurationFilePath)
becomes
Line 603 
            config.readfp(codecs.open(configurationFilePath, 'r', 'utf-8-sig'))
`

**Comment by dbamaster** [comment]
This is not the solution you are looking for, but an option could be to set the trace flags using T-SQL:

```sql
DBCC TRACEON (9481, -1);
DBCC TRACEON (2371, -1);
```
I will try to reproduce the problem on my laptop, I'm interested to see the results when using macOS.

Cheers!

**Comment by abair34** [comment]
Thanks for the reply @dbamaster!  

I tested the alternative you provided on the 2019 latest image and I was able to get it to succeed.  

>RUN /opt/mssql/bin/mssql-conf traceflag 9481 on && \
    /opt/mssql/bin/mssql-conf traceflag 2371 on

For anyone else who is interested in the alternative solution you provided I was able to get it to work in my tests against the 2019 CU4 image .  I had to modify it slightly as the latest images are non root and it appears the command runs as mssql and fails because mssql-conf tries to _sudo_ but sudo is not installed on the image.  By switching to root to run the command and then switching back to the mssql user everything ran successfully and the trace flags were enabled as desired.

```
USER root
RUN /opt/mssql/bin/mssql-conf traceflag 9481 on && \
    /opt/mssql/bin/mssql-conf traceflag 1234 on
USER mssql

```



I have noticed in Microsoft's official documentation they have suggested running echo commands to configure sql on docker containers.  

https://docs.microsoft.com/en-us/sql/linux/usage-and-diagnostic-data-configuration-for-sql-server-linux?view=sql-server-ver15

> Add an mssql.conf file with the lines [telemetry] and customerfeedback = false in the host directory:
> 
```
echo '[telemetry]' >> <host directory>/mssql.conf
echo 'customerfeedback = false' >> <host directory>/mssql.conf
```
> 

Honestly, I would rather see your solution on that page.  

I was surprised to see that their solution to configuring a SQL instance was stringing together multiple echo commands to build a config file.

I also like having the option of pushing a known config file to the container.  However based on my experience that will be troublesome for developers if they don't pay close attention to the file encoding.  

It would be great to see MS update the official doc's to use your solution and also update mssql-conf so it can handle UTF-8 with or without BOM signatures.

Thanks for your help @dbamaster!

**Comment by dbamaster** [comment]
@abair34 great news!! 👍

That's correct, I focused on the run command and totally forgot to mention the root vs mssql thing in the Dockerfile. Thanks for pointing out that! 

>Honestly, I would rather see your solution on that page.  

Ha! I might need to get in touch the MSFT SQL Server team then 😁

>I was surprised to see that their solution to configuring a SQL instance was stringing together >multiple echo commands to build a config file.

Looking at the official documentation in question, I noticed something. This example is focused on making the mssql.conf file changes, once the container is created/started (*with volumes*). Rather than providing an example of a custom build (*Dockerfile*).

I will have my head around to think in a "dynamic" way to do it using an existing config file. This is really good stuff, that I would like to share and blog about it 👌🤓

In the meantime, this issue can be closed 😀

Cheers!
